numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -322,94 +322,99 @@ PyArray_BoolConverter(PyObject *object, npy_bool *val)
     return NPY_SUCCEED;
 }
 
-/*NUMPY_API
- * Convert object to endian
- */
-NPY_NO_EXPORT int
-PyArray_ByteorderConverter(PyObject *obj, char *endian)
+static int
+string_converter_helper(
+    PyObject *object,
+    void *out,
+    int (*str_func)(char const*, Py_ssize_t, void*),
+    char const *name,
+    char const *message)
 {
-    char *str;
-    PyObject *tmp = NULL;
-
-    if (PyUnicode_Check(obj)) {
-        obj = tmp = PyUnicode_AsASCIIString(obj);
+    /* allow bytes for compatibility */
+    PyObject *str_object = NULL;
+    if (PyBytes_Check(object)) {
+        str_object = PyUnicode_FromEncodedObject(object, NULL, NULL);
+        if (str_object == NULL) {
+            PyErr_Format(PyExc_ValueError,
+                "%s %s (got %R)", name, message, object);
+            return NPY_FAIL;
+        }
     }
-
-    *endian = NPY_SWAP;
-    str = PyBytes_AsString(obj);
-    if (!str) {
-        Py_XDECREF(tmp);
+    else if (PyUnicode_Check(object)) {
+        str_object = object;
+        Py_INCREF(str_object);
+    }
+    else {
+        PyErr_Format(PyExc_TypeError,
+            "%s must be str, not %s", name, Py_TYPE(object)->tp_name);
         return NPY_FAIL;
     }
-    if (strlen(str) < 1) {
-        PyErr_SetString(PyExc_ValueError,
-                        "Byteorder string must be at least length 1");
-        Py_XDECREF(tmp);
+
+    Py_ssize_t length;
+    char const *str = PyUnicode_AsUTF8AndSize(str_object, &length);
+    if (str == NULL) {
+        Py_DECREF(str_object);
         return NPY_FAIL;
     }
-    *endian = str[0];
-    if (str[0] != NPY_BIG && str[0] != NPY_LITTLE
-        && str[0] != NPY_NATIVE && str[0] != NPY_IGNORE) {
-        if (str[0] == 'b' || str[0] == 'B') {
-            *endian = NPY_BIG;
-        }
-        else if (str[0] == 'l' || str[0] == 'L') {
-            *endian = NPY_LITTLE;
-        }
-        else if (str[0] == 'n' || str[0] == 'N') {
-            *endian = NPY_NATIVE;
-        }
-        else if (str[0] == 'i' || str[0] == 'I') {
-            *endian = NPY_IGNORE;
-        }
-        else if (str[0] == 's' || str[0] == 'S') {
-            *endian = NPY_SWAP;
-        }
-        else {
+
+    int ret = str_func(str, length, out);
+    Py_DECREF(str_object);
+    if (ret < 0) {
             PyErr_Format(PyExc_ValueError,
-                         "%s is an unrecognized byteorder",
-                         str);
-            Py_XDECREF(tmp);
-            return NPY_FAIL;
-        }
+                "%s %s (got %R)", name, message, object);
```

Comment:
No, because  `str` is a `char const *`!

Reply:
```suggestion
                "%s() %s (got %R)", name, message, object);
```
Before, I forget it on the real pass. I personally like the brackets and python typically puts them. Since we don't access `str` here, we are fine with references yeah.

Reply:
> No, because `str` is a `char const *`!

my bad, not sure what i was looking at :)

Reply:
@seberg, would vote against parens because this is not a function or class name